docs/spec/auth/subkeys_proposal.md

```diff
@@ -0,0 +1,156 @@
+
+```golang
+type DailyFeeSpend struct {
+  Address              sdk.AccAddress
+  SubKeyIndex          uint
+  FeeSpent             sdk.Coins
+}
+```
+
+When a transaction using a SubKey has paid its fee, into this queue's store we insert a new entry.
+- The key is `sdk.FormatTimeBytes(ctx.BlockTime.Add(time.Day * 1)`
+- The value is the Amino marshalled `DailyFeeSpend`
+
+Finally in the EndBlocker we iterate over the all the DailyFeeSpends in the queue which are expired (more than a day old).  We prune them from state, and deduct their FeeSpent amount from the corresponding SubKeyMetadata.
+
+```
+store := ctx.Store(spentFeeQueueStoreKey)
+iterator := sdk.Iterator(nil, sdk.FormatTimeBytes(ctx.BlockTime)) // pseudocode abstract this as a slice of DailyFeeSpend
+
+for _, dailyFeeSpend := range iterator {
+  acc := accountKeeper.GetAccount(dailyFeeSpend.Address)
+  acc.SubKeys[dailyFeeSpend.SubKeyIndex - 1].DailyFeeUsed -= dailyFeeSpend.FeeSpent
+  store.Delete(dailyFeeSpend)
+}
+```
+
+## Msgs
+
+In order to add, revoke, and update SubKeys, we would need to add some new `Msg` types to the auth module.
+
+We start by defining an `MsgAddSubKey` struct:
+
+```golang
+type MsgAddSubKey struct {
+  Address              sdk.AccAddress
+  PubKey               sdk.AccPubKey
+  PermissionedRoutes   []string
+  DailyFeeAllowance    sdk.Coins
+}
+```
+
+The handler of this Msg will create a new SubKeyMetadata struct using the data from this Msg, and append it to the end of the SubKeys slice of the Account associated with the Address.
+
+```
+acc := accountKeeper.GetAccount(msg.Address)
+
+acc.SubKeys := append(acc.SubKeys, SubKeyMetadata{
+  PubKey               msg.PubKey
+  PermissionedRoutes   msg.PermissionedRoutes
+  DailyFeeAllowance    msg.DailyFeeAllowance
+  DailyFeeUsed         sdk.Coins{}
+})
+
+accountKeeper.SetAccount(acc)
+```
+
+To make it easier to reason about the security of a SubKey, we do not allow you to add or remove permissions from a SubKey. If you want a SubKey with new permissions, you should create a new SubKey.  We do however allow you to update the allowance of a specific subkey. This can be useful because minimum transaction fee costs may fluctuate over time.
+
+To update a SubKey, we create a new `MsgUpdateSubKeyAllowance`
+
+```golang
+type MsgMsgUpdateSubKeyAllowanceAddSubKey struct {
+  Address              sdk.AccAddress
+  SubKeyIndex          uint
+  DailyFeeAllowance    sdk.Coins
+}
+```
+
+The handler of this message will update the SubKey at the index of `SubKeyIndex` in the account for `Address`.
+
+```
+if SubKeyIndex == 0 {
+  return Err
+}
+
+acc := accountKeeper.GetAccount(msg.Address)
+
+acc.SubKeys[msg.SubKeyIndex - 1].DailyFeeAllowance = msg.DailyFeeAllowance
+
+accountKeeper.SetAccount(acc)
+```
+
+We do not allow for the deletion of a SubKey, because removing it from state would mess up the indexing of subkeys. Instead, if you wish to revoke a SubKey, set its DailyFeeAllowance to 0.
```

Comment:
I don't know if this really works, because proposers can just accept zero-fee transactions. I think we need a stronger revocation that prevents any future signed transactions by the subkey.

Reply:
You're right.  Forgot MinFees aren't enforced in consensus.  Will fix.